File: 8vim/src/main/java/inc/flide/vim8/geometry/GeometricUtilities.java
package inc.flide.vim8.geometry;

import android.graphics.PointF;

import inc.flide.vim8.structures.Direction;

public final class GeometricUtilities {

    private GeometricUtilities() {
    }

    public static double getSquaredDistanceBetweenPoints(PointF a, PointF b) {
        double xSquare = Math.pow(a.x - b.x, 2);
        double ySquare = Math.pow(a.y - b.y, 2);
        return Math.abs(xSquare + ySquare);
    }

    public static Direction getBaseQuadrant(int continuousQuadrantValue) {
        int result;
        // Calculate result with modulus operator
        result = continuousQuadrantValue % 4;
        // Fix zero truncation
        if (result < 0) {
            result += 4;
        }
        try {
            return Direction.values()[result];
        } catch (Exception e) {
            return null;
        }
    }
}

File: 8vim/src/main/java/inc/flide/vim8/geometry/GeometricUtilities.kt
package inc.flide.vim8.geometry

import android.graphics.PointF
import kotlin.math.*

object GeometricUtilities {
    fun getSquaredDistanceBetweenPoints(a: PointF, b: PointF): Double {
        val xSquare = (a.x - b.x).toDouble().pow(2.0)
        val ySquare = (a.y - b.y).toDouble().pow(2.0)
        return abs(xSquare + ySquare)
    }

    fun findPointSpecifiedDistanceAwayInGivenDirection(startingPoint: PointF, directionalAngleInDegree: Double, length: Double): PointF {
        val directionalAngleInRadians = Math.toRadians(directionalAngleInDegree)
        val x = (startingPoint.x + length * cos(directionalAngleInRadians)).toFloat().roundToInt()
        val y = (startingPoint.y + length * sin(directionalAngleInRadians)).toFloat().roundToInt()
        return PointF(x.toFloat(), y.toFloat())
    }
}